src/erlcloud_ec2.erl

```diff
@@ -1441,27 +1538,131 @@ extract_block_device_mapping_status(Node) ->
 
-%%
-%%
--spec(describe_instance_status/1 :: (string()) -> proplist()).
-describe_instance_status(InstanceID) ->
-    describe_instance_status([{"InstanceId", InstanceID}], [], default_config()).
+-type instance_id() :: string().
+-type ec2_param_list() :: [{string(),string()}].
 
--spec(describe_instance_status/2 :: (proplist(), filter_list()) -> proplist()).
-describe_instance_status(Params, Filter) ->
-    describe_instance_status(Params, Filter, default_config()).
+-define(SPOT_INSTANCE_STATUS_MR_MIN, 5).
+-define(SPOT_INSTANCE_STATUS_MR_MAX, 1000).
 
--spec(describe_instance_status/3 :: (proplist(), filter_list(), aws_config()) -> proplist()).
-describe_instance_status(Params, Filter, Config) ->
+-spec describe_instance_status() -> ec2_normal_describe_return().
+describe_instance_status() ->
+    describe_instance_status([], [], default_config()).
+
+-spec describe_instance_status(instance_id()) -> ec2_normal_describe_return();
+                              (aws_config()) -> ec2_normal_describe_return();
+                              (ec2_max_result()) -> ec2_paged_describe_return().
+describe_instance_status(InstanceID)
+    when is_list(InstanceID) ->
+    describe_instance_status([{"InstanceId", InstanceID}], [], default_config());
+describe_instance_status(Config)
+    when is_record(Config, aws_config) ->
+    describe_instance_status([], [], Config);
+describe_instance_status(MaxResults)
+    when is_integer(MaxResults) ->
+    describe_instance_status([], [], MaxResults, undefined, default_config()).
+
+-spec describe_instance_status(ec2_param_list(), filter_list()) -> ec2_normal_describe_return();
+                              (ec2_param_list(), aws_config()) -> ec2_normal_describe_return();
+                              (ec2_param_list(), ec2_max_result()) -> ec2_paged_describe_return();
+                              (ec2_max_result(), aws_config()) -> ec2_paged_describe_return();
+                              (ec2_max_result(), ec2_token()) -> ec2_paged_describe_return().
+describe_instance_status(Params, Filter)
+    when is_list(Params), is_list(Filter) ->
+    describe_instance_status(Params, Filter, default_config());
+describe_instance_status(Params, Config)
+    when is_list(Params), is_record(Config, aws_config) ->
+    describe_instance_status(Params, [], Config);
+describe_instance_status(Params, MaxResults)
+    when is_list(Params), is_integer(MaxResults) ->
+    describe_instance_status(Params, [], MaxResults, undefined, default_config());
+describe_instance_status(MaxResults, Config)
+    when is_integer(MaxResults), is_record(Config, aws_config) ->
+    describe_instance_status([], [], MaxResults, undefined, Config);
+describe_instance_status(MaxResults, NextToken)
+    when is_integer(MaxResults), is_list(NextToken) ->
+    describe_instance_status([], [], MaxResults, NextToken, default_config()).
+
+-spec describe_instance_status(ec2_param_list(), filter_list(), aws_config()) -> ec2_normal_describe_return();
```

Comment:
I am sorry, I don't believe that it the best way to design api. There was clear `describe_instance_status(proplist(), filter_list(), aws_config())` where you always had config as a 3rd argument. But after these changes we basically can pass `ec2_max_result() | ec2_token()` instead (which implies default config?).

To me it looks like a mess, to be honest. Can we leave it as is and add 4th arg for tokens? Or if some of the args are mutually exclusive reflect it.